### ISSUE TITLE
[$500] Error shows up whenever tried to go back using keyboard

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#34474]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.24-7
**Reproducible in staging?:** yes
**Reproducible in production?:** no
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @shubham1206agra 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1705158369037609

## Action Performed:
1. Open App
2. Open FAB > Assign task
3. Type something in title, and go to next screen.
4. With keyboard, try to go back or press on any menuitem.

## Expected Result: 
No validation error should show up, and page should go back on using back button.

## Actual Result: 
Validation error should shows up, and page doesn't go back.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/47e005c7-b98d-42bb-9564-f2017b2d6927

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016f34652481906d29</li>
        <li>Upwork Job ID: 1746196695910436864</li>
        <li>Last Price Increase: 2024-01-20</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Contributor | 28115099</li>
<li>shubham1206agra | Contributor | 28117382</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sakluger</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sakluger</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sakluger</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The task validation error shows up when we try to select other pressable using keyboard.

### What is the root cause of that problem?
This is another regression from https://github.com/Expensify/App/pull/33055 where it [updates the `useActiveElementRole` to use ref instead of state](https://github.com/Expensify/App/pull/33055/files#diff-592e09f097641e6728d742d07d7729ce0e5af8951cc3e4503b99b0688df6754dL1-L32), so it won't re-render the components that use it when the role changes.

`useActiveElementRole` is used to disable e ENTER shortcut for a button, in our case the task confirm button.
https://github.com/Expensify/App/blob/753ce92ea9a1858dd2d576eeff692df527ea3ecb/src/components/Button/index.tsx#L129

### What changes do you think we should make in order to solve the problem?

Use state back in `useActiveElementRole`

**Comment by Julesssss** [comment]
@shubham1206agra suggested an improvement to a linked issue [here](https://github.com/Expensify/App/issues/28916#issuecomment-1890864528) that resolves this issue.  I'm going to assign them to handle this fix and make the suggested improvement.

**Comment by shubham1206agra** [comment]
Copying the comment here

I have tried to further optimize @hurali97's [proposal](https://github.com/Expensify/App/issues/28916#issuecomment-1853920181) here https://github.com/Expensify/App/pull/34480.

I tried to do 2 main things 

1. Instead of creating 2 new listener for `focusin`, and `focusout` for every use of `useActiveElementRole`, we are going to create context of `activeRole`, and will listen only once for events inside the Provider, which will be supplied to App root component directly. The aim was to remove the redundancy of listeners, as we know every button will have these listeners, and there are almost no places on App where we don't have a button.
2. Instead of using the above context for every button rendered, we have used `pressOnEnter` prop to conditionally render `KeyboardShortcutComponent`, which will help in less re-render on changing of `activeRole`.

Now we use state for storing `activeRole` instead of ref cause this caused a regression https://github.com/Expensify/App/issues/34474, and restoring to using state directly causing some performance lag.

**Comment by bernhardoj** [comment]
Using state back won't introduce the performance lag in KeyboardShortcutComponent (as it renders nothing), but possibly in `BaseSelectionList`.

**Comment by shubham1206agra** [comment]
> Using state back won't introduce the performance lag in KeyboardShortcutComponent (as it renders nothing), but possibly in `BaseSelectionList`. `BaseSelectionList` should follow the same pattern as `Button` to avoid any unnecessary re-render of the whole component.

But it tries to run state update logic for every button on changing of `activeRole`, and will takes some resources to perform these unnecessary calculations. I have checked this, and RAM usage does spike a little. 

> `BaseSelectionList` should follow the same pattern as `Button` to avoid any unnecessary re-render of the whole component.

I can also do the same changes once these changes are finalized.

**Comment by Julesssss** [comment]
Yeah, we'll still do C+ review as usual here. I'll edit my comment.

**Comment by sakluger** [comment]
@0xmiroslav can you review today please? As mentioned, this is blocking another issue.

**Comment by shubham1206agra** [comment]
@0xmiroslav The above issue seems out of scope of this issue. This issue is about the listener not working properly.

**Comment by shubham1206agra** [comment]
@0xmiroslav Upon my investigation, I came to the problem that goBack is executing twice very quickly. Please see the below trace of `goBack`.

<img width="587" alt="Screenshot 2024-01-19 at 6 11 53 PM" src="https://github.com/Expensify/App/assets/58412969/b3b5d8e1-94d0-4d48-b424-ec232c713806">

<img width="587" alt="Screenshot 2024-01-19 at 6 12 47 PM" src="https://github.com/Expensify/App/assets/58412969/5532b8d8-f903-4dd3-b983-89cf94fd5210">

You can observe that the problem is occurring due to onPress is getting executed from 2 different places (both in `react-native-web` [we don't have control over that]) in spite of `singleExecution` call on onPress handler. I am once again saying that this issue has different RCA, and should have different scope. 

[This solution](https://github.com/Expensify/App/issues/34474#issuecomment-1890565006) will not solve it correctly at all. On the other hand, it will introduce some memory bloating as we are subscribing to every event whether it is active or not, and just doing return early.

**Comment by situchan** [comment]
@shubham1206agra I am confused: `Now we use state for storing activeRole instead of ref.` - Did you state this as bug (current behavior) or part of your solution?

**Comment by shubham1206agra** [comment]
> @shubham1206agra I am confused: `Now we use state for storing activeRole instead of ref.` - Did you state this as bug or part of your solution? Because you said `We cannot use state for storing activeRole directly instead of ref as this is causing some performance lag.` in RCA

As part of solution.

**Comment by situchan** [comment]
`useActiveElementRole`:
This is general **hook** and the components where hooked should re-render when its value changes.
If using ref, they're not re-rendered so we should handle **sync** always in the components where that hook is used. So I don't follow @dukenv0307's solution as it loses hook ability.

@bernhardoj proposed first to use state instead of ref. But I consider Option 2 below substantial improvement in @shubham1206agra's proposal.

> Option 2. Instead of using the above context for every button rendered, we have used `pressOnEnter` prop to conditionally render `KeyboardShortcutComponent`, which will help in less re-render on changing of `activeRole`.

Finally, I recommend @shubham1206agra's [proposal](https://github.com/Expensify/App/issues/34474#issuecomment-1892288533).
🎀 👀 🎀 C+ reviewed

**Comment by Julesssss** [comment]
> @Julesssss Can you start payment process here?

Hey @sakluger, can you handle this please? PR was deployed 2 weeks ago, looks like our automation failed somehow.

**Comment by sakluger** [comment]
Sorry for the delay! Thanks @mallenexpensify for adding the daily label.

Summarizing payment on this issue:

Contributor: @shubham1206agra $500, sent offer via Upwork (https://www.upwork.com/nx/wm/offer/101367666)
Contributor+: @situchan $500, paid via Upwork

**Comment by sakluger** [comment]
Hey @shubham1206agra, please let me know once you've accepted the offer. Thanks!

**Comment by sakluger** [comment]
@shubham1206agra what's the latest?

**Comment by sakluger** [comment]
Sounds good @shubham1206agra. I've withdrawn the Upwork offer.

The payment summary is [here](https://github.com/Expensify/App/issues/34474#issuecomment-1995977439).

**Comment by sakluger** [comment]
@shubham1206agra fyi I don't always see comments on already closed GH issues. Feel free to bump me in Slack sooner if payment is due but the issue is closed.

**Comment by sakluger** [comment]
@shubham1206agra here is the Upwork offer (https://www.upwork.com/nx/wm/offer/102288115) since it sounds like we are not going to pay via manual request.

**Comment by sakluger** [comment]
Thanks! All paid 🙇
